C7/Map/MapViewCamera.cs

```diff
@@ -0,0 +1,116 @@
+			checkWorldWrap();
+		}
+
+		public void setZoom(float factor) {
+			zoomFactor = Mathf.Clamp(factor, minZoom, maxZoom);
+			Zoom = Vector2.One * zoomFactor;
+			checkWorldWrap();
+		}
+
+		public void setPosition(Vector2 position) {
+			Position = position;
+			checkWorldWrap();
+		}
+		private float wrapEdgeMargin = 0;
+		float wrapRightEdge {get => map.worldEdgeRight - wrapEdgeMargin; }
+		float wrapLeftEdge {get => map.worldEdgeLeft + wrapEdgeMargin; }
+		private bool enteringRightWrap(Rect2 v) => hwrap != HorizontalWrapState.Right && v.End.X >= wrapRightEdge;
+		private bool enteringLeftWrap(Rect2 v) => hwrap != HorizontalWrapState.Left && v.Position.X <= wrapLeftEdge;
+		private bool atEdgeOfRightWrap(Rect2 v) => hwrap == HorizontalWrapState.Right && v.End.X >= wrapRightEdge + map.pixelWidth;
+		private bool atEdgeOfLeftWrap(Rect2 v) => hwrap == HorizontalWrapState.Left && v.Position.X <= wrapLeftEdge - map.pixelWidth;
+		private HorizontalWrapState currentHwrap(Rect2 v) {
+			return v.Position.X <= wrapLeftEdge ? HorizontalWrapState.Left : (v.End.X >= wrapRightEdge ? HorizontalWrapState.Right : HorizontalWrapState.None);
+		}
+
+		// checkWorldWrap determines if the camera is about to spill over the world map and will:
+		// - move the second "wrap" tilemap to the appropriate edge
+		//   to give the illusion of true wrapping tilemap
+		// - teleport the camera one world-width to the left or right when
+		//   only the "wrap" tilemap is in view
+
+		private void checkWorldWrap() {
+			if (map is null || !map.wrapHorizontally) {
+				// TODO: for maps that do not wrap horizontally restrict movement
+				return;
+			}
+			Rect2 visibleWorld = getVisibleWorld();
+			if (enteringRightWrap(visibleWorld)) {
+				map.setHorizontalWrap(HorizontalWrapState.Right);
+			} else if (enteringLeftWrap(visibleWorld)) {
+				map.setHorizontalWrap(HorizontalWrapState.Left);
+			}
+			if (atEdgeOfRightWrap(visibleWorld)) {
+				Translate(Vector2.Left * map.pixelWidth);
+			} else if (atEdgeOfLeftWrap(visibleWorld)) {
+				Translate(Vector2.Right * map.pixelWidth);
+			}
+			hwrap = currentHwrap(visibleWorld);
+		}
+
+		public override void _UnhandledInput(InputEvent @event) {
+			switch (@event) {
+				case InputEventMouseMotion mouseDrag when mouseDrag.ButtonMask == MouseButtonMask.Left:
+					setPosition(Position - mouseDrag.Relative / Zoom);
+					break;
+				case InputEventMagnifyGesture magnifyGesture:
+					scaleZoom(magnifyGesture.Factor);
+					break;
+			}
+		}
+
+		public Rect2 getVisibleWorld() {
+			Transform2D vpToGlobal = (GetViewport().GlobalCanvasTransform * GetCanvasTransform()).AffineInverse();
+			return vpToGlobal * GetViewportRect();
+		}
+
+		public void centerOnTile(Tile tile, MapView map) {
+			Vector2 target = map.tileToLocal(tile);
+			setPosition(target);
+		}
+
+		public bool isTileInView(Tile tile, MapView map) {
+			Rect2 visible = getVisibleWorld();
+			Vector2 target = map.tileToLocal(tile);
+			float size = 30;
```

Comment:
What's this number? Can it be derived from something, and it not made a constant somewhere? I think we'd discussed having a class to hold all map related constants.